File: 1375.bulb-switcher-iii/1375.bulb-switcher-iii.py
#
# @lc app=leetcode id=1375 lang=python3
#
# [1375] Bulb Switcher III
#
from typing import List
# @lc code=start


class Solution:
    def numTimesAllBlue(self, light: List[int]) -> int:
        result, lastBlue, maxLight, cache = 0, 0, 0, {}
        for i in range(len(light)):
            l = light[i]
            cache[l] = True
            if l > maxLight:
                maxLight = l
            if lastBlue == l-1:
                lastBlue = l
            else:
                continue
            allBlue = True
            for j in range(lastBlue+1, maxLight+1):
                if cache.__contains__(j) and cache[j]:
                    lastBlue = j
                else:
                    allBlue = False
                    break
            if allBlue:
                result += 1
        return result

# @lc code=end

File: 1375.bulb-switcher-iii/1375.bulb-switcher-iii.go
/*
 * @lc app=leetcode id=1375 lang=golang
 *
 * [1375] Bulb Switcher III
 */

// @lc code=start
func numTimesAllBlue(light []int) int {
	result := 0
	lastBlue, maxLight := 0, 0
	cache := make(map[int]bool, 0)
	for i := 0; i < len(light); i++ {
		if light[i] > maxLight {
			maxLight = light[i]
		}
		cache[light[i]] = true
		if lastBlue+1 == light[i] {
			lastBlue = light[i]
		}
		if lastBlue != light[i] {
			continue
		}
		allBlue := true
		for i := lastBlue + 1; i <= maxLight; i++ {
			if b, ok := cache[i]; !ok || !b {
				allBlue = false
				break
			}
			lastBlue = i
		}
		if allBlue {
			result++
		}
	}
	return result
}

// @lc code=end
